### ISSUE TITLE
Support reading arrays of unknown size from JSON via push_back() on an array

### ISSUE THREAD

**Issue by buddha314** [chapel-lang/chapel#7582]
Bruce said this would be a good idea.

```
  record Recipe {
    var cuisine: string,
        id: int,
        ingredients: list(string);
  }
  record CookBook {
    var recipes: list(Recipe);
}

  proc loadTrainingData() {
    var cookBook = new CookBook();
    try {
      var f = open(data, iomode.r);
      var r = f.reader();
      r.readf("%jt\n", cookBook);
    } catch {
      halt();
    }
    return cookBook;
}
```

```
{"recipes":
[
  {
    "id": 10259,
    "cuisine": "greek",
    "ingredients": [
      "romaine lettuce",
      "black olives",
      "grape tomatoes",
      "garlic",
      "pepper",
      "purple onion",
      "seasoning",
      "garbanzo beans",
      "feta cheese crumbles"
    ]
  },
  {
    "id": 25693,
    "cuisine": "southern_us",
    "ingredients": [
      "plain flour",
      "ground pepper",
      "salt",
      "tomatoes",
      "ground black pepper",
      "thyme",
      "eggs",
      "green tomatoes",
      "yellow corn meal",
      "milk",
      "vegetable oil"
    ]
  }
]
```

**Comment by bradcray** [comment]
Hi @mppf:  @benharsh and I were trying to help @buddha314 with this today and found ourselves wondering whether reading arrays of unknown size from JSON files could be handled by using push_back() on an array.  At a high level it seems as though it should work, and would be desirable, but I don't know much about JSON in general or our implementation of reading JSON in particular.  The workarounds we were coming up with were... not pretty (e.g., using lists and then sloshing those lists into arrays so we could do real computation with them).  Thanks.

**Comment by mppf** [comment]
@bradcray - the reason we didn't support this was due to discomfort with the resulting language design. It's certainly the case that we could use push_back to implement resizing 1-D arrays, and that we could describe the feature as "Resize on read happens with any array that we can use push_back on". But fundamentally we still have the language design question. Are you saying that you now think that we should support resize-on-read when reading with JSON into a 1-D array?

We should discuss the design issue here: #5053.

As far as *actually* using push_back goes, I don't think that would be possible without some refactoring since chpl_serialReadWriteRectangular takes in a DefaultRectangularArray but push_back is implemented in terms of _array. The shortest implementation path is probably to make it work similarly to push_back (which I had done earlier, but it's effectively commented out pending a design decision, and some other code adjustments have made simply uncommenting it no longer work).

Put another way - I'd be happy to adjust the implementation once we know what it's supposed to do.

**Comment by benharsh** [comment]
Is #7524 a duplicate?

**Comment by mppf** [comment]
Yes, this issue,  #7524, and  #5053 are all really about the same thing, but with slightly different takes on the solution.

**Comment by buddha314** [comment]
don't harsh my ticket, ben.
